README.md
# Rust bindings for EGL

<table><tr>
  <td><a href="https://docs.rs/khronos-egl">Documentation</a></td>
  <td><a href="https://crates.io/crates/khronos-egl">Crate informations</a></td>
  <td><a href="https://github.com/timothee-haudebourg/khronos-egl">Repository</a></td>
</tr></table>

This crate provides a binding for the Khronos EGL 1.5 API.
It was originally a fork of the [egl](https://crates.io/crates/egl) crate,
which is left unmaintained.

## Usage

You can access the EGL API using an [`Instance`](https://docs.rs/khronos-egl/latest/khronos-egl/struct.Instance.html)
object defined by either statically linking with `libEGL.so.1` at compile time,
or dynamically loading the EGL library at runtime.

### Static linking

You must enable static linking using the `static` feature in your `Cargo.toml`:
```toml
khronos-egl = { version = ..., features = ["static"] }
```

This will add a dependency to the [`pkg-config`](https://crates.io/crates/pkg-config) crate,
necessary to find the EGL library at compile time.
Here is a simple example showing how to use this library to create an EGL context when static linking is enabled.

```rust
extern crate khronos_egl as egl;

fn main() -> Result<(), egl::Error> {
	// Create an EGL API instance.
	// The `egl::Static` API implementation is only available when the `static` feature is enabled.
	let egl = egl::Instance::new(egl::Static);

	let wayland_display = wayland_client::Display::connect_to_env().expect("unable to connect to the wayland server");
	let display = egl.get_display(wayland_display.get_display_ptr() as *mut std::ffi::c_void).unwrap();
	egl.initialize(display)?;

	let attributes = [
		egl::RED_SIZE, 8,
		egl::GREEN_SIZE, 8,
		egl::BLUE_SIZE, 8,
		egl::NONE
	];

	let config = egl.choose_first_config(display, &attributes)?.expect("unable to find an appropriate ELG configuration");

	let context_attributes = [
		egl::CONTEXT_MAJOR_VERSION, 4,
		egl::CONTEXT_MINOR_VERSION, 0,
		egl::CONTEXT_OPENGL_PROFILE_MASK, egl::CONTEXT_OPENGL_CORE_PROFILE_BIT,
		egl::NONE
	];

	egl.create_context(display, config, None, &context_attributes);

	Ok(())
}
```

The creation of a `Display` instance is not detailed here since it depends on your display server.
It is created using the `get_display` function with a pointer to the display server connection handle.
For instance, if you are using the [wayland-client](https://crates.io/crates/wayland-client) crate,
you can get this pointer using the `Display::get_display_ptr` method.

#### Static API Instance

It may be bothering in some applications to pass the `Instance` to every fonction that needs to call the EGL API.
One workaround would be to define a static `Instance`,
which should be possible to define at compile time using static linking.
However this is not yet supported by the stable `rustc` compiler.
With the nightly compiler,
you can combine the `nightly` and `static` features so that this crate
can provide a static `Instance`, called `API` that can then be accessed everywhere.

```rust
use egl::API as egl;
```

### Dynamic Linking

Dynamic linking allows your application to accept multiple versions of EGL and be more flexible.
You must enable dynamic linking using the `dynamic` feature in your `Cargo.toml`:
```toml
khronos-egl = { version = ..., features = ["dynamic"] }
```

This will add a dependency to the [`libloading`](https://crates.io/crates/libloading) crate,
necessary to find the EGL library at runtime.
You can then load the EGL API into a `Instance<Dynamic<libloading::Library>>` as follows:

```rust
let lib = libloading::Library::new("libEGL.so.1").expect("unable to find libEGL.so.1");
let egl = unsafe { egl::DynamicInstance::<egl::EGL1_4>::load_required_from(lib).expect("unable to load libEGL.so.1") };
```

Here, `egl::EGL1_4` is used to specify what is the minimum required version of EGL that must be provided by `libEGL.so.1`.
This will return a `DynamicInstance<egl::EGL1_4>`, however in that case where `libEGL.so.1` provides a more recent version of EGL,
you can still upcast ths instance to provide version specific features:
```rust
match egl.upcast::<egl::EGL1_5>() {
	Some(egl1_5) => {
		// do something with EGL 1.5
	}
	None => {
		// do something with EGL 1.4 instead.
	}
};
```

### NixOS

A `shell.nix` file is present for nix users to build the crate easily.
Just enter a new nix shell using the given configuration file,
and `cargo build` should work.
If you want to run the tests and examples you will need to use `shell-wayland.nix` instead
that will also load wayland since most of them depend on it.

## Testing

Most test and examples most be compiled with the `static` feature.

## Troubleshooting

### Static Linking with OpenGL ES

When using OpenGL ES with `khronos-egl` with the `static` feature,
it is necessary to place a dummy extern at the top of your application which links libEGL first, then GLESv1/2.
This is because libEGL provides symbols required by GLESv1/2.
Here's how to work around this:

```rust
#[link(name = "EGL")]
#[link(name = "GLESv2")]
extern {}
```

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

If the original `egl` crate was licensed only under the Apache 2.0 license,
I believe I have made enough breaking changes so that no relevant code from the
original code remains and the rest can be relicensed.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unpublished]
### Changed
- `copy_buffers` is now unsafe.
- `get_display` is now unsafe.
- `get_platform_display` is now unsafe.
- `create_platform_window_surface` is now unsafe.
- `create_platform_pixmap_surface` is now unsafe.
### Fixed
- Fix `choose_config` & `get_config` undefined behavior when the input `configs`
  vector is empty. (Fixes [#21](https://github.com/timothee-haudebourg/khronos-egl/issues/21)).
- Fix Windows build (Fixes [#23](https://github.com/timothee-haudebourg/khronos-egl/pull/23)).

## [5.0.0]
### Changed
- Upgrade dependency `libloading`: ^0.7 -> ^0.8.

## [4.1.0]
### Changed
- `load_required` and `load` now trying to load `libEGL.so.1` or `libEGL.so`.

## [4.0.0]
### Added
- `no-pkg-config` feature.
### Changed
- Upgrade dependency `libloading`: ^0.6 -> ^0.7.
### Removed
- `nightly` feature hich is no longer needed since `const_fn` is stabilized.

## [3.0.2]
### Changed
- One Linux, use the `RTLD_NODELETE` when loading the EGL library in `load_required_from_filename` and `load_from_filename`.

## [3.0.1]
### Changed
- Load `libEGL.so.1` by default instead of `libEGL.so`.

## [3.0.0]
### Changed
- Impl `Debug` for `Static`, `Dynamic` and `Instance`.
- Add a `DynamicInstance` type alias for `Instance<Dynamic<libloading::Library>>` with helper functions.
- Precise version selection.
- Dynamic cast between versions with `Dynamic::load`, `Dynamic::load_required` and the `Upcast`/`Downcast` traits.
- `DynamicInstance::downcast` and `IDynamicInstance::upcast`.

## [3.0.0-beta]
### Changed
- Removed the `khronos` dependency.
- Dynamic linking: Add the `Api` trait and the `Instance` struct along with the `static` and `dynamic` features.
- The dependency to `pkg-config` is now optional, only required by the `static` feature.
- Add an optional dependency to `libloading`, only required by the `dynamic` feature.

## [2.2.0]
### Added
- Fix #9: new function `get_config_count` to get the number of available frame buffer configurations.

## [2.1.1]
### Changed
- Upgrade dependency `gl`: ^0.11 -> ^0.14
- Upgrade dependency `wayland-client`: ^0.23 -> ^0.25
- Upgrade dependency `wayland-protocols`: ^0.23 -> ^0.25
- Upgrade dependency `wayland-egl`: ^0.23 -> ^0.25

## [2.1.0]
### Changed
- Fix #3: accept `Option<Display>` instead of `Display` in `query_string`.
- More flexible dependencies versions.


examples/wayland.rs
extern crate gl;
extern crate khronos_egl as egl;
extern crate wayland_client;
extern crate wayland_egl;
extern crate wayland_protocols;

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::ptr;
use std::ffi::CStr;
use gl::types::{GLint, GLuint, GLchar, GLenum, GLboolean, GLvoid};

use wayland_client::{
	protocol::{wl_compositor::WlCompositor, wl_surface::WlSurface},
	DispatchData, Display, EventQueue, Main
};

use wayland_protocols::xdg_shell::client::{
	xdg_surface::{self, XdgSurface},
	xdg_wm_base::{self, XdgWmBase},
};

fn process_xdg_event(xdg: Main<XdgWmBase>, event: xdg_wm_base::Event, _dd: DispatchData) {
	use xdg_wm_base::Event::*;

	match event {
		Ping { serial } => xdg.pong(serial),
		_ => (),
	}
}

struct DisplayConnection {
	display: Display,
	event_queue: EventQueue,
	compositor: Main<WlCompositor>,
	xdg: Main<XdgWmBase>,
}

fn setup_wayland() -> DisplayConnection {
	let display =
		wayland_client::Display::connect_to_env().expect("unable to connect to the wayland server");
	let mut event_queue = display.create_event_queue();
	let attached_display = display.clone().attach(event_queue.token());

	let globals = wayland_client::GlobalManager::new(&attached_display);

	// Roundtrip to retrieve the globals list
	event_queue
		.sync_roundtrip(&mut (), |_, _, _| unreachable!())
		.unwrap();

	// Get the compositor.
	let compositor: Main<WlCompositor> = globals.instantiate_exact(1).unwrap();

	// Xdg protocol.
	let xdg: Main<XdgWmBase> = globals.instantiate_exact(1).unwrap();
	xdg.quick_assign(process_xdg_event);

	DisplayConnection {
		display,
		event_queue,
		compositor,
		xdg,
	}
}

fn setup_egl(display: &Display) -> egl::Display {
	let egl_display = egl::get_display(display.get_display_ptr() as *mut std::ffi::c_void).unwrap();
	egl::initialize(egl_display).unwrap();

	egl_display
}

fn create_context(display: egl::Display) -> (egl::Context, egl::Config) {
	let attributes = [
		egl::RED_SIZE,
		8,
		egl::GREEN_SIZE,
		8,
		egl::BLUE_SIZE,
		8,
		egl::NONE,
	];

	let config = egl::choose_first_config(display, &attributes)
		.expect("unable to choose an EGL configuration")
		.expect("no EGL configuration found");

	let context_attributes = [
		egl::CONTEXT_MAJOR_VERSION,
		4,
		egl::CONTEXT_MINOR_VERSION,
		0,
		egl::CONTEXT_OPENGL_PROFILE_MASK,
		egl::CONTEXT_OPENGL_CORE_PROFILE_BIT,
		egl::NONE,
	];

	let context = egl::create_context(display, config, None, &context_attributes)
		.expect("unable to create an EGL context");

	(context, config)
}

struct Surface {
	handle: Main<WlSurface>,
	initialized: AtomicBool
}

fn create_surface(
	ctx: &DisplayConnection,
	egl_display: egl::Display,
	egl_context: egl::Context,
	egl_config: egl::Config,
	width: i32,
	height: i32,
) -> Arc<Surface> {
	let wl_surface = ctx.compositor.create_surface();
	let xdg_surface = ctx.xdg.get_xdg_surface(&wl_surface);

	let xdg_toplevel = xdg_surface.get_toplevel();
	xdg_toplevel.set_app_id("khronos-egl-test".to_string());
	xdg_toplevel.set_title("Test".to_string());

	wl_surface.commit();
	ctx.display.flush().unwrap();

	let surface = Arc::new(Surface {
		handle: wl_surface,
		initialized: AtomicBool::new(false)
	});

	let weak_surface = Arc::downgrade(&surface);

	xdg_surface.quick_assign(move |xdg_surface: Main<XdgSurface>, event: xdg_surface::Event, _dd: DispatchData| {
		use xdg_surface::Event::*;

		match event {
			Configure { serial } => {
				if let Some(surface) = weak_surface.upgrade() {
					if !surface.initialized.swap(true, Ordering::Relaxed) {
						let wl_egl_surface = wayland_egl::WlEglSurface::new(&surface.handle, width, height);

						let egl_surface = unsafe {
							egl::create_window_surface(
								egl_display,
								egl_config,
								wl_egl_surface.ptr() as egl::NativeWindowType,
								None,
							)
							.expect("unable to create an EGL surface")
						};

						egl::make_current(egl_display, Some(egl_surface), Some(egl_surface), Some(egl_context))
							.expect("unable to bind the context");

						render();

						egl::swap_buffers(egl_display, egl_surface)
							.expect("unable to post the surface content");

						xdg_surface.ack_configure(serial);
					}
				}
			},
			_ => (),
		}
	});

	surface
}

fn main() {
	// Setup Open GL.
	egl::bind_api(egl::OPENGL_API)
		.expect("unable to select OpenGL API");
	gl::load_with(|name| egl::get_proc_address(name).unwrap() as *const std::ffi::c_void);

	// Setup the Wayland client.
	let mut ctx = setup_wayland();

	// Setup EGL.
	let egl_display = setup_egl(&ctx.display);
	let (egl_context, egl_config) = create_context(egl_display);

	// Create a surface.
	// Note that it must be kept alive to the end of execution.
	let _surface = create_surface(&ctx, egl_display, egl_context, egl_config, 800, 600);

	loop {
		ctx.event_queue
			.dispatch(&mut (), |_, _, _| { /* we ignore unfiltered messages */ })
			.unwrap();
	}
}

const VERTEX: &'static [GLint; 8] = &[
	-1, -1,
	1, -1,
	1, 1,
	-1, 1
];

const INDEXES: &'static [GLuint; 4] = &[
	0, 1, 2, 3
];

const VERTEX_SHADER: &[u8] = b"#version 400
in vec2 position;

void main() {
	gl_Position = vec4(position, 0.0f, 1.0f);
}
\0";

const FRAGMENT_SHADER: &[u8] = b"#version 400
out vec4 color;

void main() {
	color = vec4(1.0f, 0.0f, 0.0f, 1.0f);
}
\0";

fn render() {
	unsafe {
		let vertex_shader = gl::CreateShader(gl::VERTEX_SHADER);
		check_gl_errors();
		let src = CStr::from_bytes_with_nul_unchecked(VERTEX_SHADER).as_ptr();
		gl::ShaderSource(vertex_shader, 1, (&[src]).as_ptr(), ptr::null());
		check_gl_errors();
		gl::CompileShader(vertex_shader);
		check_shader_status(vertex_shader);

		let fragment_shader = gl::CreateShader(gl::FRAGMENT_SHADER);
		check_gl_errors();
		let src = CStr::from_bytes_with_nul_unchecked(FRAGMENT_SHADER).as_ptr();
		gl::ShaderSource(fragment_shader, 1, (&[src]).as_ptr(), ptr::null());
		check_gl_errors();
		gl::CompileShader(fragment_shader);
		check_shader_status(fragment_shader);

		let program = gl::CreateProgram();
		check_gl_errors();
		gl::AttachShader(program, vertex_shader);
		check_gl_errors();
		gl::AttachShader(program, fragment_shader);
		check_gl_errors();
		gl::LinkProgram(program);
		check_gl_errors();
		gl::UseProgram(program);
		check_gl_errors();

		let mut buffer = 0;
		gl::GenBuffers(1, &mut buffer);
		check_gl_errors();
		gl::BindBuffer(gl::ARRAY_BUFFER, buffer);
		check_gl_errors();
		gl::BufferData(
			gl::ARRAY_BUFFER,
			8 * 4,
			VERTEX.as_ptr() as *const std::ffi::c_void,
			gl::STATIC_DRAW
		);
		check_gl_errors();

		let mut vertex_input = 0;
		gl::GenVertexArrays(1, &mut vertex_input);
		check_gl_errors();
		gl::BindVertexArray(vertex_input);
		check_gl_errors();
		gl::EnableVertexAttribArray(0);
		check_gl_errors();
		gl::VertexAttribPointer(
			0, 2, gl::INT, gl::FALSE as GLboolean, 0, 0 as *const GLvoid
		);
		check_gl_errors();

		let mut indexes = 0;
		gl::GenBuffers(1, &mut indexes);
		check_gl_errors();
		gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, indexes);
		check_gl_errors();
		gl::BufferData(
			gl::ELEMENT_ARRAY_BUFFER,
			4 * 4,
			INDEXES.as_ptr() as *const std::ffi::c_void,
			gl::STATIC_DRAW
		);
		check_gl_errors();

		gl::DrawElements(gl::TRIANGLE_FAN, 4, gl::UNSIGNED_INT, std::ptr::null());
		check_gl_errors();
	}
}

fn format_error(e: GLenum) -> &'static str {
	 match e {
		  gl::NO_ERROR => "No error",
		  gl::INVALID_ENUM => "Invalid enum",
		  gl::INVALID_VALUE => "Invalid value",
		  gl::INVALID_OPERATION => "Invalid operation",
		  gl::INVALID_FRAMEBUFFER_OPERATION => "Invalid framebuffer operation",
		  gl::OUT_OF_MEMORY => "Out of memory",
		  gl::STACK_UNDERFLOW => "Stack underflow",
		  gl::STACK_OVERFLOW => "Stack overflow",
		  _ => "Unknown error"
	 }
}

pub fn check_gl_errors() {
	 unsafe {
		  match gl::GetError() {
				gl::NO_ERROR => (),
				e => {
					panic!("OpenGL error: {}", format_error(e))
				}
		  }
	 }
}

unsafe fn check_shader_status(shader: GLuint) {
	let mut status = gl::FALSE as GLint;
	gl::GetShaderiv(shader, gl::COMPILE_STATUS, &mut status);
	if status != (gl::TRUE as GLint) {
		let mut len = 0;
		gl::GetProgramiv(shader, gl::INFO_LOG_LENGTH, &mut len);
		if len > 0 {
			let mut buf = Vec::with_capacity(len as usize);
			buf.set_len((len as usize) - 1); // subtract 1 to skip the trailing null character
			gl::GetProgramInfoLog(
				shader,
				len,
				ptr::null_mut(),
				buf.as_mut_ptr() as *mut GLchar,
			);

			let log = String::from_utf8(buf).unwrap();
			eprintln!("shader compilation log:\n{}", log);
		}

		panic!("shader compilation failed");
	}
}


examples/load-minimal.rs
extern crate gl;
extern crate khronos_egl as egl;
extern crate libloading;

use egl::{EGL1_2, EGL1_4};
use std::sync::Arc;

fn main() {
	let minimal_egl =
		unsafe { Arc::new(egl::DynamicInstance::load().expect("unable to load libEGL.so.1")) };

	println!("EGL version is {}", minimal_egl.version());

	// Select the rendering API.
	if let Some(egl1_2) = minimal_egl.upcast::<EGL1_2>() {
		println!("selecting API");
		egl1_2
			.bind_api(egl::OPENGL_API)
			.expect("unable to select OpenGL API");
	}

	// Setup Open GL.
	gl::load_with(|name| minimal_egl.get_proc_address(name).unwrap() as *const std::ffi::c_void);

	match minimal_egl.upcast::<EGL1_4>() {
		Some(egl1_4) => foo_with_1_4(egl1_4),
		None => foo_without_1_4(&minimal_egl),
	}
}

fn foo_with_1_4<V: egl::api::EGL1_4>(_egl: &egl::Instance<V>) {
	println!("with 1.4");
	// do something that requires at least EGL 1.4.
}

fn foo_without_1_4<V>(_egl: &egl::Instance<V>) {
	println!("without 1.4");
	// do something without any specific EGL version (other that 1.0).
}


examples/wayland-dynamic.rs
extern crate gl;
extern crate khronos_egl as egl;
extern crate libloading;
extern crate wayland_client;
extern crate wayland_egl;
extern crate wayland_protocols;

use gl::types::{GLboolean, GLchar, GLenum, GLint, GLuint, GLvoid};
use std::ffi::CStr;
use std::ptr;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use wayland_client::{
	protocol::{wl_compositor::WlCompositor, wl_surface::WlSurface},
	DispatchData, Display, EventQueue, Main,
};

use wayland_protocols::xdg_shell::client::{
	xdg_surface::{self, XdgSurface},
	xdg_wm_base::{self, XdgWmBase},
};

fn process_xdg_event(xdg: Main<XdgWmBase>, event: xdg_wm_base::Event, _dd: DispatchData) {
	use xdg_wm_base::Event::*;

	match event {
		Ping { serial } => xdg.pong(serial),
		_ => (),
	}
}

struct DisplayConnection {
	display: Display,
	event_queue: EventQueue,
	compositor: Main<WlCompositor>,
	xdg: Main<XdgWmBase>,
}

fn setup_wayland() -> DisplayConnection {
	let display =
		wayland_client::Display::connect_to_env().expect("unable to connect to the wayland server");
	let mut event_queue = display.create_event_queue();
	let attached_display = display.clone().attach(event_queue.token());

	let globals = wayland_client::GlobalManager::new(&attached_display);

	// Roundtrip to retrieve the globals list
	event_queue
		.sync_roundtrip(&mut (), |_, _, _| unreachable!())
		.unwrap();

	// Get the compositor.
	let compositor: Main<WlCompositor> = globals.instantiate_exact(1).unwrap();

	// Xdg protocol.
	let xdg: Main<XdgWmBase> = globals.instantiate_exact(1).unwrap();
	xdg.quick_assign(process_xdg_event);

	DisplayConnection {
		display,
		event_queue,
		compositor,
		xdg,
	}
}

fn setup_egl(egl: &egl::DynamicInstance, display: &Display) -> egl::Display {
	let egl_display = egl
		.get_display(display.get_display_ptr() as *mut std::ffi::c_void)
		.unwrap();
	egl.initialize(egl_display).unwrap();

	egl_display
}

fn create_context(
	egl: &egl::DynamicInstance,
	display: egl::Display,
) -> (egl::Context, egl::Config) {
	let attributes = [
		egl::RED_SIZE,
		8,
		egl::GREEN_SIZE,
		8,
		egl::BLUE_SIZE,
		8,
		egl::NONE,
	];

	let config = egl
		.choose_first_config(display, &attributes)
		.expect("unable to choose an EGL configuration")
		.expect("no EGL configuration found");

	let context_attributes = [
		egl::CONTEXT_MAJOR_VERSION,
		4,
		egl::CONTEXT_MINOR_VERSION,
		0,
		egl::CONTEXT_OPENGL_PROFILE_MASK,
		egl::CONTEXT_OPENGL_CORE_PROFILE_BIT,
		egl::NONE,
	];

	let context = egl
		.create_context(display, config, None, &context_attributes)
		.expect("unable to create an EGL context");

	(context, config)
}

struct Surface {
	handle: Main<WlSurface>,
	initialized: AtomicBool,
}

fn create_surface(
	egl: &Arc<egl::DynamicInstance>,
	ctx: &DisplayConnection,
	egl_display: egl::Display,
	egl_context: egl::Context,
	egl_config: egl::Config,
	width: i32,
	height: i32,
) -> Arc<Surface> {
	let wl_surface = ctx.compositor.create_surface();
	let xdg_surface = ctx.xdg.get_xdg_surface(&wl_surface);

	let xdg_toplevel = xdg_surface.get_toplevel();
	xdg_toplevel.set_app_id("khronos-egl-test".to_string());
	xdg_toplevel.set_title("Test".to_string());

	wl_surface.commit();
	ctx.display.flush().unwrap();

	let surface = Arc::new(Surface {
		handle: wl_surface,
		initialized: AtomicBool::new(false),
	});

	let weak_surface = Arc::downgrade(&surface);

	let egl = egl.clone();
	xdg_surface.quick_assign(
		move |xdg_surface: Main<XdgSurface>, event: xdg_surface::Event, _dd: DispatchData| {
			use xdg_surface::Event::*;

			match event {
				Configure { serial } => {
					if let Some(surface) = weak_surface.upgrade() {
						if !surface.initialized.swap(true, Ordering::Relaxed) {
							let wl_egl_surface =
								wayland_egl::WlEglSurface::new(&surface.handle, width, height);

							let egl_surface = unsafe {
								egl.create_window_surface(
									egl_display,
									egl_config,
									wl_egl_surface.ptr() as egl::NativeWindowType,
									None,
								)
								.expect("unable to create an EGL surface")
							};

							egl.make_current(
								egl_display,
								Some(egl_surface),
								Some(egl_surface),
								Some(egl_context),
							)
							.expect("unable to bind the context");

							render();

							egl.swap_buffers(egl_display, egl_surface)
								.expect("unable to post the surface content");

							xdg_surface.ack_configure(serial);
						}
					}
				}
				_ => (),
			}
		},
	);

	surface
}

fn main() {
	let egl = unsafe {
		Arc::new(
			egl::DynamicInstance::<egl::EGL1_5>::load_required()
				.expect("unable to load libEGL.so.1"),
		)
	};

	// Setup Open GL.
	egl.bind_api(egl::OPENGL_API)
		.expect("unable to select OpenGL API");
	gl::load_with(|name| egl.get_proc_address(name).unwrap() as *const std::ffi::c_void);

	// Setup the Wayland client.
	let mut ctx = setup_wayland();

	// Setup EGL.
	let egl_display = setup_egl(&egl, &ctx.display);
	let (egl_context, egl_config) = create_context(&egl, egl_display);

	// Create a surface.
	// Note that it must be kept alive to the end of execution.
	let _surface = create_surface(&egl, &ctx, egl_display, egl_context, egl_config, 800, 600);

	loop {
		ctx.event_queue
			.dispatch(&mut (), |_, _, _| { /* we ignore unfiltered messages */ })
			.unwrap();
	}
}

const VERTEX: &'static [GLint; 8] = &[-1, -1, 1, -1, 1, 1, -1, 1];

const INDEXES: &'static [GLuint; 4] = &[0, 1, 2, 3];

const VERTEX_SHADER: &[u8] = b"#version 400
in vec2 position;

void main() {
	gl_Position = vec4(position, 0.0f, 1.0f);
}
\0";

const FRAGMENT_SHADER: &[u8] = b"#version 400
out vec4 color;

void main() {
	color = vec4(1.0f, 0.0f, 0.0f, 1.0f);
}
\0";

fn render() {
	unsafe {
		let vertex_shader = gl::CreateShader(gl::VERTEX_SHADER);
		check_gl_errors();
		let src = CStr::from_bytes_with_nul_unchecked(VERTEX_SHADER).as_ptr();
		gl::ShaderSource(vertex_shader, 1, (&[src]).as_ptr(), ptr::null());
		check_gl_errors();
		gl::CompileShader(vertex_shader);
		check_shader_status(vertex_shader);

		let fragment_shader = gl::CreateShader(gl::FRAGMENT_SHADER);
		check_gl_errors();
		let src = CStr::from_bytes_with_nul_unchecked(FRAGMENT_SHADER).as_ptr();
		gl::ShaderSource(fragment_shader, 1, (&[src]).as_ptr(), ptr::null());
		check_gl_errors();
		gl::CompileShader(fragment_shader);
		check_shader_status(fragment_shader);

		let program = gl::CreateProgram();
		check_gl_errors();
		gl::AttachShader(program, vertex_shader);
		check_gl_errors();
		gl::AttachShader(program, fragment_shader);
		check_gl_errors();
		gl::LinkProgram(program);
		check_gl_errors();
		gl::UseProgram(program);
		check_gl_errors();

		let mut buffer = 0;
		gl::GenBuffers(1, &mut buffer);
		check_gl_errors();
		gl::BindBuffer(gl::ARRAY_BUFFER, buffer);
		check_gl_errors();
		gl::BufferData(
			gl::ARRAY_BUFFER,
			8 * 4,
			VERTEX.as_ptr() as *const std::ffi::c_void,
			gl::STATIC_DRAW,
		);
		check_gl_errors();

		let mut vertex_input = 0;
		gl::GenVertexArrays(1, &mut vertex_input);
		check_gl_errors();
		gl::BindVertexArray(vertex_input);
		check_gl_errors();
		gl::EnableVertexAttribArray(0);
		check_gl_errors();
		gl::VertexAttribPointer(0, 2, gl::INT, gl::FALSE as GLboolean, 0, 0 as *const GLvoid);
		check_gl_errors();

		let mut indexes = 0;
		gl::GenBuffers(1, &mut indexes);
		check_gl_errors();
		gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, indexes);
		check_gl_errors();
		gl::BufferData(
			gl::ELEMENT_ARRAY_BUFFER,
			4 * 4,
			INDEXES.as_ptr() as *const std::ffi::c_void,
			gl::STATIC_DRAW,
		);
		check_gl_errors();

		gl::DrawElements(gl::TRIANGLE_FAN, 4, gl::UNSIGNED_INT, std::ptr::null());
		check_gl_errors();
	}
}

fn format_error(e: GLenum) -> &'static str {
	match e {
		gl::NO_ERROR => "No error",
		gl::INVALID_ENUM => "Invalid enum",
		gl::INVALID_VALUE => "Invalid value",
		gl::INVALID_OPERATION => "Invalid operation",
		gl::INVALID_FRAMEBUFFER_OPERATION => "Invalid framebuffer operation",
		gl::OUT_OF_MEMORY => "Out of memory",
		gl::STACK_UNDERFLOW => "Stack underflow",
		gl::STACK_OVERFLOW => "Stack overflow",
		_ => "Unknown error",
	}
}

pub fn check_gl_errors() {
	unsafe {
		match gl::GetError() {
			gl::NO_ERROR => (),
			e => {
				panic!("OpenGL error: {}", format_error(e))
			}
		}
	}
}

unsafe fn check_shader_status(shader: GLuint) {
	let mut status = gl::FALSE as GLint;
	gl::GetShaderiv(shader, gl::COMPILE_STATUS, &mut status);
	if status != (gl::TRUE as GLint) {
		let mut len = 0;
		gl::GetProgramiv(shader, gl::INFO_LOG_LENGTH, &mut len);
		if len > 0 {
			let mut buf = Vec::with_capacity(len as usize);
			buf.set_len((len as usize) - 1); // subtract 1 to skip the trailing null character
			gl::GetProgramInfoLog(
				shader,
				len,
				ptr::null_mut(),
				buf.as_mut_ptr() as *mut GLchar,
			);

			let log = String::from_utf8(buf).unwrap();
			eprintln!("shader compilation log:\n{}", log);
		}

		panic!("shader compilation failed");
	}
}
